File: src/test/java/cube8540/oauth/authentication/credentials/oauth/token/application/CompositionAuthorizationCodeServiceTest.java
package cube8540.oauth.authentication.credentials.oauth.token.application;

import cube8540.oauth.authentication.credentials.oauth.security.AuthorizationRequest;
import cube8540.oauth.authentication.credentials.oauth.token.domain.AuthorizationCodeGenerator;
import cube8540.oauth.authentication.credentials.oauth.token.domain.AuthorizationCodeRepository;
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2AuthorizationCode;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;

import java.util.Optional;

import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.CLIENT_ID;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.RAW_AUTHORIZATION_CODE;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.REDIRECT_URI;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.SCOPES;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.USERNAME;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.makeAuthorizationCode;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.makeAuthorizationCodeRepository;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.makeAuthorizationRequest;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.makeCodeGenerator;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.makeEmptyAuthorizationCodeRepository;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@DisplayName("기본 인가 코드 서비스 테스트")
class CompositionAuthorizationCodeServiceTest {

    @Test
    @DisplayName("저장소에 등록 되지 않은 인가 코드 검색")
    void consumeNotRegisteredAuthorizationCodeInRepository() {
        AuthorizationCodeRepository repository = makeEmptyAuthorizationCodeRepository();
        CompositionAuthorizationCodeService service = new CompositionAuthorizationCodeService(repository);

        Optional<OAuth2AuthorizationCode> result = service.consume(RAW_AUTHORIZATION_CODE);
        assertEquals(Optional.empty(), result);
        verify(repository, never()).delete(any());
    }

    @Test
    @DisplayName("인가 코드 검색")
    void consumeAuthorizationCode() {
        OAuth2AuthorizationCode code = makeAuthorizationCode();
        AuthorizationCodeRepository repository = makeAuthorizationCodeRepository(RAW_AUTHORIZATION_CODE, code);
        CompositionAuthorizationCodeService service = new CompositionAuthorizationCodeService(repository);

        Optional<OAuth2AuthorizationCode> result = service.consume(RAW_AUTHORIZATION_CODE);
        assertTrue(result.isPresent());
        assertEquals(code, result.get());
        verify(repository, times(1)).delete(code);
    }

    @Test
    @DisplayName("새 인가 코드 생성")
    void generateNewAuthorizationCode() {
        ArgumentCaptor<OAuth2AuthorizationCode> codeArgumentCaptor = ArgumentCaptor.forClass(OAuth2AuthorizationCode.class);
        AuthorizationCodeRepository repository = makeEmptyAuthorizationCodeRepository();
        AuthorizationRequest request = makeAuthorizationRequest();
        AuthorizationCodeGenerator generator = makeCodeGenerator(RAW_AUTHORIZATION_CODE);
        CompositionAuthorizationCodeService service = new CompositionAuthorizationCodeService(repository);

        service.setCodeGenerator(generator);

        service.generateNewAuthorizationCode(request);
        verify(repository, times(1)).save(codeArgumentCaptor.capture());
        assertEquals(RAW_AUTHORIZATION_CODE, codeArgumentCaptor.getValue().getCode());
        assertEquals(CLIENT_ID, codeArgumentCaptor.getValue().getClientId());
        assertEquals(SCOPES, codeArgumentCaptor.getValue().getApprovedScopes());
        assertEquals(REDIRECT_URI, codeArgumentCaptor.getValue().getRedirectURI());
        assertEquals(USERNAME, codeArgumentCaptor.getValue().getUsername());
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/oauth/token/application/CompositionAuthorizationCodeServiceTest.kt
package cube8540.oauth.authentication.oauth.token.application

import com.nimbusds.oauth2.sdk.pkce.CodeChallenge
import com.nimbusds.oauth2.sdk.pkce.CodeChallengeMethod
import com.nimbusds.oauth2.sdk.pkce.CodeVerifier
import cube8540.oauth.authentication.oauth.security.AuthorizationRequest
import cube8540.oauth.authentication.oauth.token.domain.AuthorizationCodeGenerator
import cube8540.oauth.authentication.oauth.token.domain.AuthorizationCodeRepository
import cube8540.oauth.authentication.oauth.token.domain.OAuth2AuthorizationCode
import cube8540.oauth.authentication.oauth.token.domain.PrincipalUsername
import cube8540.oauth.authentication.security.AuthorityCode
import io.mockk.*
import java.net.URI
import java.util.Optional
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class CompositionAuthorizationCodeServiceTest {

    private val generator: AuthorizationCodeGenerator = mockk()
    private val repository: AuthorizationCodeRepository = mockk(relaxed = true) {
        every { save(any()) } returnsArgument 0
    }

    private val service = CompositionAuthorizationCodeService(repository)

    init {
        service.codeGenerator = generator
    }

    @Nested
    inner class ConsumeTest {

        @Test
        fun `authorization code not registered`() {
            every { repository.findById("authorizationCode") } returns Optional.empty()

            val result = service.consume("authorizationCode")
            assertThat(result).isNull()
            verify(exactly = 0) { repository.delete(any()) }
        }

        @Test
        fun `authorization code successful`() {
            val authorizationCode: OAuth2AuthorizationCode = mockk()

            every { repository.findById("authorizationCode") } returns Optional.of(authorizationCode)

            val result = service.consume("authorizationCode")
            assertThat(result).isEqualTo(authorizationCode)
            verify(exactly = 1) { repository.delete(authorizationCode) }
        }
    }

    @Nested
    inner class CodeGenerateTest {

        @Test
        fun `generate new code`() {
            val codeArgumentCaptor = slot<OAuth2AuthorizationCode>()
            val authorizationRequest: AuthorizationRequest = mockk {
                every { clientId } returns "clientId"
                every { requestScopes } returns setOf("scope-1", "scope-2", "scope-3")
                every { redirectUri } returns URI.create("http://localhost")
                every { username } returns "username"
                every { codeChallengeMethod } returns CodeChallengeMethod.S256
                every { codeChallenge } returns CodeChallenge.compute(CodeChallengeMethod.S256, CodeVerifier("FP7Am8xqMbyTCBgSYiTVuVkVv8ffScYCt2wali8JVC8"))
            }

            every { generator.generate() } returns "authorizationCode"
            every { repository.save(capture(codeArgumentCaptor)) } returnsArgument 0

            service.generateNewAuthorizationCode(authorizationRequest)
            assertThat(codeArgumentCaptor.isCaptured).isTrue
            assertThat(codeArgumentCaptor.captured.code).isEqualTo("authorizationCode")
            assertThat(codeArgumentCaptor.captured.username).isEqualTo(PrincipalUsername("username"))
            assertThat(codeArgumentCaptor.captured.approvedScopes).isEqualTo(
                setOf(AuthorityCode("scope-1"), AuthorityCode("scope-2"), AuthorityCode("scope-3")))
            assertThat(codeArgumentCaptor.captured.redirectURI).isEqualTo(URI.create("http://localhost"))
            assertThat(codeArgumentCaptor.captured.codeChallengeMethod).isEqualTo(CodeChallengeMethod.S256)
            assertThat(codeArgumentCaptor.captured.codeChallenge).isEqualTo(CodeChallenge.compute(CodeChallengeMethod.S256, CodeVerifier("FP7Am8xqMbyTCBgSYiTVuVkVv8ffScYCt2wali8JVC8")))
        }
    }

    @AfterEach
    fun afterClear() {
        clearAllMocks()
    }
}